backend-rust/src/graphql_api/account_metrics.rs

```diff
@@ -0,0 +1,110 @@
+use std::sync::Arc;
+
+use async_graphql::{Context, Object, SimpleObject};
+use sqlx::postgres::types::PgInterval;
+
+use super::{get_pool, ApiError, ApiResult, DateTime, MetricsPeriod, TimeSpan};
+
+#[derive(SimpleObject)]
+struct AccountMetrics {
+    /// Total number of accounts created (all time).
+    last_cumulative_accounts_created: i64,
+
+    /// Total number of accounts created in requested period.
+    accounts_created: i64,
+
+    buckets: AccountMetricsBuckets,
+}
+
+#[derive(SimpleObject)]
+struct AccountMetricsBuckets {
+    /// The width (time interval) of each bucket.
+    bucket_width: TimeSpan,
+
+    /// Start of the bucket time period. Intended x-axis value.
+    #[graphql(name = "x_Time")]
+    x_time: Vec<DateTime>,
+
+    /// Total number of accounts created (all time) at the end of the bucket
+    /// period. Intended y-axis value.
+    #[graphql(name = "y_LastCumulativeAccountsCreated")]
+    y_last_cumulative_accounts_created: Vec<i64>,
+
+    /// Number of accounts created within bucket time period. Intended y-axis
+    /// value.
+    #[graphql(name = "y_AccountsCreated")]
+    y_accounts_created: Vec<i64>,
+}
+
+#[derive(Default)]
+pub(crate) struct AccountMetricsQuery;
+
+#[Object]
+impl AccountMetricsQuery {
+    async fn account_metrics(
+        &self,
+        ctx: &Context<'_>,
+        period: MetricsPeriod,
+    ) -> ApiResult<AccountMetrics> {
+        let pool = get_pool(ctx)?;
+
+        let last_cumulative_accounts_created =
+            sqlx::query_scalar!("SELECT MAX(index) FROM accounts").fetch_one(pool).await?.unwrap();
```

Comment:
**Suggestion:** Avoid panicking.

Since this service will only read data from the database, it can't really do much damage, so I think crashing in the case of something being off is not ideal. It will take down every other query being processed. If we introduce some bug, which can trigger this, it would allow an attacker to easily do a denial-of-service attack.

Instead, I suggest returning an Internal error to the user, which we would then monitor for.

Reply:
>It will take down every other query being processed.

That's not true - axum catches panics so that only that single request that panics is affected. It will not crash the entire web server, it will just drop the HTTP connection ungracefully. Not too dissimilar to a single thread panicking while other threads are unaffected.

I can definitely understand that you'd want to avoid panics if it would affect other requests or the server as a whole, but it does not. It could not be used as part of a DOS attack.

Reply:
Reporting an error instead of panicking (even the database should never be corrupted) is better because it makes debugging easier in my opinion if that ever happens.

Reply:
I don't disagree with that in general, but in this case, this error will essentially never happen and therefore an unwrap is fine I think. The panic will point to the line of code in question and should make it obvious what happened so I don't think, in this particular case, that it hurts debugging.

Reply:
The error case can happen if e.g. the indexer has not started yet (meaning no data in the database). This is an error and we should be easily debug it from the front-end to report to devOps that they need to start the indexing service.

Reply:
Can't find any information on what axum does when there is a panic, but shutting down the connection ungracefully is not nice either.
For better error message to the user and codebase consistency, it would be good to do the same as we do elsewhere and handle the error.

Reply:
I'm not sure where exactly it's documented but that is how it works. If we don't like dropping the connection, we could use the [catch panic middleware layer](https://docs.rs/tower-http/latest/tower_http/catch_panic/index.html) that turns it into a 500 response instead. But I think that's outside the scope of this PR.

Reply:
I see that axum spawns a task per request and the panicking is handled by the tokio runtime then, so you are right in it not being a vulnerability.
I would still prefer consistency and a proper error messages for the user.

Reply:
Alright, I opted to coalesce the max instead, to make it so there is no error and the endpoint will simply report 0 in the case of no data, which seems more correct than giving an error.